File: Leetcode/Reorder_Data_in_Log_Files.cpp
// Problem : letter log를 digit log 보다 앞에오게하고, letter log들은 뒷부분 문자열로 정렬하고, 문자가 똑같으면 id로 정렬.
// Time : O(NlogN), Space : O(N)

class Solution {
public:
    vector<string> reorderLogFiles(vector<string>& logs) {
        stable_sort(logs.begin(), logs.end(), [](string l, string r) {
            int i = l.find(" ");
            int j = r.find(" ");
            if (!isdigit(l[i+1]) && !isdigit(r[j+1])) {
                string l_sub = l.substr(i+1);
                string r_sub = r.substr(j+1);
                return (l_sub != r_sub) ? (l_sub < r_sub) : (l.substr(0, i) < r.substr(0, j));
            }
            else {
                if(isdigit(l[i+1]))
                    return false;
                return true;
            }
        });
        return logs;
    }
};

File: Leetcode/Reorder_Data_in_Log_Files.py
# Problem : let 으로 시작하는 원소를 dig 보다 앞에오게하고, let들은 뒷부분 문자열로 정렬하고, 문자가 똑같으면 id로 정렬.
# Time : O(NlogN), Space : O(N)


class Solution:
    def sort(self, e):
        i, log = e.split(" ", 1)
        if log[0].isalpha():
            return (0, log, i)
        else:
            return (1, None, None)

    def reorderLogFiles(self, logs: List[str]) -> List[str]:
        return sorted(logs, key=self.sort)
